feat: switch DNS resolver to Unbound from systemd-resolved

alexvojproc opened the pull request:
This PR switches the local DNS stub resolver to Unbound, due to systemd-resolved's unreliable DNSSEC validation and broader concerns. See https://github.com/secureblue/secureblue/issues/1174.

Changelog:
- Remove systemd-resolved and install dnsconfd-unbound
  - dnsconfd implements a dbus interface for NetworkManager and applies DNS settings dynamically to Unbound
  - Configured NetworkManager to use dnsconfd dbus interface
- Harden unbound.service and configure Unbound
  - Run unprivileged, no capabilities, no setuid/setgid/chroot
  - Use systemd network and UNIX sockets
  - Lock down other syscalls, namespaces, filesystem
  - Fedora bug opened, hopefully will upstream a subset
  - Add non-default hardening and performance configuration
- Harden dnsconfd.service
  - Will try to upstream
- Migrate existing configuration to NetworkManager with secureblue-migrate-dns.service
  - Seamless for all users of the bash `ujust dns-selector`
- Refactor `ujust dns-selector` and rewrite in Python using sandboxing framework
  - Pre-defined servers from bash now stored in JSON at /usr/share/secureblue
- Rewrite `ujust audit-secureblue` `audit_dns()`
  - Checks for global DNS, DNSSEC validation, Trivalent DoT
  - Wrapper around new `ujust dns-selector status`
- Update i18n for `ujust audit-secureblue`
  - Some translatable strings have changed
- Add workaround for unapplied dnsconfd SELinux type
  - Upstream has a bug in v1.7.2 that causes dnsconfd to run unconfined by SELinux.
  - Fixed in v1.7.3, pinged maintainer so now [on Bodhi](https://bodhi.fedoraproject.org/updates/FEDORA-2025-2eb1c80e01), autopushes in 1-2 days?

Not included in this PR:
- Per-connection DNS, to simplify configuration of DNS with VPNs and work/public networks
  - Also VPN-enlightened `ujust audit-secureblue`
- Zero-trust DNS even in Anaconda by setting kargs, so no leaks (possible from F43)
- Investigation needed for GNOME's overly strict GUI DNS validation
- DNS allowlisting

alexvojproc committed:
feat(securedns): replace systemd-resolved with dnsconfd-unbound

HastD commented on files/po/audit_secureblue.pot:
```diff
@@ -1,23 +1,14 @@
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
+# SOME DESCRIPTIVE TITLE.
```
Looks like you accidentally overwrote the copyright and license header.

HastD commented on files/system/usr/lib/systemd/system/dnsconfd.service.d/secureblue.conf:
```diff
@@ -0,0 +1,40 @@
+RestrictNamespaces=yes
+PrivateDevices=yes
+SystemCallArchitectures=native
+SystemCallFilter=@system-service @network-io @io-event @signal
```
This is redundant: the syscall sets `@network-io`, `@io-event`, and `@signal`  are included in `@system-service`.

HastD commented on files/system/usr/lib/systemd/system/unbound.service.d/secureblue.conf:
```diff
@@ -0,0 +1,86 @@
+StateDirectory=unbound
+# Upstream was ~@clock @cpu-emulation @debug @keyring @module mount @obsolete
+# @resources
+SystemCallFilter=@system-service @network-io @io-event @signal
```
Again, this is redundant, `@system-service` includes all of these. However, shouldn't there also be a deny-list (`SystemCallFilter=~...`) to avoid weakening the syscall filter relative to upstream? For example, based on your comment above, `@keyring` is denied in the upstream service unit, but it's part of `@system-service`, so it'll be allowed as written here.

alexvojproc replied:
aha, I missed that! thank you

alexvojproc replied:
I've kept this line, but have also re-added the line from upstream beneath this. Is that ok? @HastD

HastD replied:
I don't see any reason to include the redundant syscall sets since they're already covered by `@system-service`.

HastD commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -357,50 +357,54 @@ def audit_chronyd():
+            _("System DNS resolution may not be secure."),
+            _("To configure this, run:"),
+            "$ ujust dns-selector",
+        ]
```
This is a single recommendation; you should call this `rec_lines` or something, join it by newlines, and append it as a single string to `recs`.

HastD commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -357,50 +357,54 @@ def audit_chronyd():
+        recs += [
+            _("You should enable local DNSSEC validation, by running:"),
+            "$ ujust dns-selector dnssec on",
+        ]
```
Likewise, this is a single recommendation.

HastD commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -357,50 +357,54 @@ def audit_chronyd():
+
+    rec_string = None if recs == [] else "\n".join(recs)
+    yield Report(
+        _("Ensuring system DNS resolution is secure"), status, warnings=warnings, recs=rec_string
```
...and then you can pass `recs` (as a list) to this argument. (The constructor accepts a list.)

HastD commented on files/system/usr/libexec/secureblue/dns_selector.py:
```diff
@@ -0,0 +1,383 @@
+DNSCONFD_CONF_FILE = Path("/etc/dnsconfd.conf")
+NM_GLOBALDNS_CONF_FILE = Path("/etc/NetworkManager/conf.d/global-dns.conf")
+TRIVALENT_POLICY_FILE = Path("/etc/trivalent/policies/managed/10-securedns-browser.json")
+SERVERS_JSON_FILE = Path("/usr/share/secureblue/secure-dns-providers.json")
```
It's good practice to annotate global constants as `Final` (e.g. `Final[Path]` in this case) so type-checkers will validate that they're not accidentally modified later on.

HastD commented on files/system/usr/libexec/secureblue/dns_selector.py:
```diff
@@ -0,0 +1,383 @@
+    """Returns the user's preference between 'yes'/'y' (True) and 'no'/'n' (False)."""
+
+    while True:
+        ans = interruptible_ask(prompt + " [y/n] ").lower()
```
Minor nitpick but this should be `.casefold()`, not `.lower()`.

HastD commented on files/system/usr/libexec/secureblue/dns_selector.py:
```diff
@@ -0,0 +1,383 @@
+            return ans in ("y", "yes")
+        print("Please enter y (yes) or n (no).")
+
+
+def ask_should_use_doh() -> bool:
+    """Returns the user's preference for Trivalent DoH enforcement (yes = true, no = false)."""
+
+    print(
+        "Would you like to enable DNS over HTTPS (DoH) in the Trivalent browser?\n"
+        "This tool already configures encrypted DNS over TLS, but DoH looks like "
+        "HTTPS traffic to outsiders, which could have privacy benefits.\n"
+        "1. Enable DoH for Trivalent (masks DNS queries);\n"
+        "2. Use the same encrypted DNS as the rest of the system in Trivalent."
+    )
```
You can make multi-line print blocks like this more readable with `textwrap.dedent`:
```py
import textwrap
print(textwrap.dedent("""
    One line of text
    Next line
    ...
    Last line
""".strip()))
```

HastD commented on files/system/usr/libexec/secureblue/dns_selector.py:
```diff
@@ -0,0 +1,383 @@
+    )
+    mode = ask_option(dnssec)
+
+    if mode == reset:
```
This can be a `match` block:
```py
match mode:
    case reset:
        # ...
    case enforce:
        # ...
    case dnssec:
        # ...
```

HastD commented on files/system/usr/libexec/secureblue/dns_selector.py:
```diff
@@ -0,0 +1,383 @@
+                    if value.strip().lower() in ("yes", "true"):
+                        print("DNSSEC: enabled")
+                    return
+    except FileNotFoundError:
```
What about a `PermissionError` or other `OSError`?

alexvojproc replied:
I've updated this to handle `(PermissionError, OSError, UnicodeDecodeError)` here and in one other place just with an additional "DNSSEC: unable to open and parse configuration" value. Is that ok?

HastD replied:
`PermissionError` is a subclass of `OSError` so you can omit it here, but other than that, looks good.

alexvojproc replied:
oops!

HastD commented on files/system/usr/libexec/secureblue/dns_selector.py:
```diff
@@ -0,0 +1,383 @@
+    dnssec_p.add_argument("state", choices=["on", "off"])
+    args = p.parse_args()
+
+    if args.cmd is None:
```
This can also use a `match` block for `args.cmd`. Note that you can have a fallback case matching on `_` for the "invalid option selected" case.

HastD commented on files/system/usr/libexec/secureblue/inner/dns.py:
```diff
@@ -0,0 +1,149 @@
+    try:
+        # We're calling system binaries with fixed arguments and no shell.
+        # nosemgrep: dangerous-subprocess-use-audit
+        subprocess.run(["/usr/bin/systemctl", "restart", "dnsconfd.service"], check=True)
```
I think a `# nosec` comment at the end of the line should suppress the Codacy warning here.

HastD reviewed (changes requested):
Thanks for your work on this. Haven't tested it out on my system, just a few minor comments from reading through the code.

RoyalOughtness commented on files/scripts/selinux/dnsconfd_bin/dnsconfd_bin.te:
```diff
@@ -0,0 +1,5 @@
+policy_module(dnsconfd_bin, 1.0.0)
+
+require {
+    type dnsconfd_exec_t;
```
this policy currently is a no-op. What is its purpose?

alexvojproc replied:
the policy exists for the `dnsconfd_bin.fc`, which correctly types /usr/bin/dnsconfd

RoyalOughtness replied:
why does it need to be typed though? what is the overall purpose of the policy?

alexvojproc replied:
the policy supplements the existing upstream dnsconfd policy (see [dnsconfd.te](https://github.com/InfrastructureServices/dnsconfd/blob/main/distribution/dnsconfd.te)). in v1.7.2 the _dnsconfd_t context_ works as intended, but the dnsconfd exec does not have _dnsconfd_exec_t_ [applied properly](https://github.com/InfrastructureServices/dnsconfd/blob/1.7.2/distribution/dnsconfd.fc). in v1.7.3, which is now hitting F42 stable in a couple of days, this is fixed. the patch SELinux 'policy' I wrote only has a stub .te, but its **dnsconfd.fc** then labels /usr/bin/dnsconfd correctly and fixes upstream's 1.7.2 bug.

alexvojproc replied:
when I wrote it, I wasn't sure how long it'd take to get the build into stable, so I wrote that override 'policy'. I think it's sensible to wait until v1.7.3 is in F42 stable, then we can revert the SELinux commit in this PR. in hindsight, I should have just done an `semanage fcontext` override, which would have been clearer.

RoyalOughtness commented on files/scripts/enablednsconfd.sh:
```diff
@@ -0,0 +1,30 @@
+rm -f /etc/resolv.conf
+touch /etc/resolv.conf
+chmod 0644 /etc/resolv.conf
+chown dnsconfd:root /etc/resolv.conf
```
why is the group `root` if the owner isn't?

alexvojproc replied:
i was imitating the default behaviour, but there's no real reason it can't be dnsconfd:dnsconfd

RoyalOughtness replied:
if that's intentional then we can keep it

RoyalOughtness commented on files/scripts/enablednsconfd.sh:
```diff
@@ -0,0 +1,30 @@
+#   org.freedesktop.resolve1, and
+# - setting the permissions of /etc/resolv.conf manually here.
+rm -f /etc/resolv.conf
+touch /etc/resolv.conf
```
this can be done more cleanly using the files module

RoyalOughtness commented on files/system/usr/lib/systemd/system/secureblue-migrate-dns.service:
```diff
@@ -0,0 +1,56 @@
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/python3 /usr/libexec/secureblue/migrate_dns.py
```
we should test this in integration tests before enabling it in prod

RoyalOughtness commented on files/scripts/enableunboundmigration.sh:
```diff
@@ -0,0 +1,19 @@
+
+set -oue pipefail
+
+systemctl enable secureblue-migrate-dns.service
```
for now should we keep this disabled by default and start it in integration tests/local VMs to test. and then we can flip it on once it's ready?

alexvojproc replied:
do you mean the whole feature? if we just leave the migration disabled, then people currently using resolved global-dns will have an unexpected security downgrade

RoyalOughtness replied:
I mean everything, but I'm confused by what you mean. If the migration is disabled, isn't everything disabled? We shouldn't be enabling anything until after the migration, right?

alexvojproc replied:
The migration service just takes the config files written by the original Bash dns-selector (servers, DoT, DNSSEC) and converts them into the NetworkManager/dnsconfd formats. It runs once post-update before the network is up. It's simple and fairly well-tested, as it's there to seamlessly transition the vast majority of users with supported configs. After updating, nobody should notice anything except "DNSSEC works now".

The rest of the PR, by masking systemd-resolved, taking over resolv.conf and installing Unbound, is doing the enabling. I could amend the PR to support co-existence if you like, such as by not disabling systemd-resolved or migrating until the user opts in. What do you think?

RoyalOughtness replied:
@alexvojproc Ah I understand now, thanks. Should we consider merging to `staging` and having a broad set of folks test in the #testing channel?

alexvojproc replied:
@RoyalOughtness sounds good to me, assuming @HastD's suggestions wrt code quality have been addressed. maybe we can test at least until v1.7.3 hits stable, so we can revert [fix(securedns): add workaround for unapplied /usr/bin/dnsconfd type](https://github.com/secureblue/secureblue/pull/1335/commits/7b1d93a853d757ed5b8486c3ae1aac803df0f9d9)?

HastD replied:
You did address all my comments, right? I don't think I had any other concerns. I do suggest making an issue to remind us to remove this SELinux policy once v1.7.3 is on stable.

RoyalOughtness commented on files/system/etc/unbound/conf.d/secureblue.conf:
```diff
@@ -0,0 +1,11 @@
+	do-daemonize: no
+
+	# Stop unbound from binding to non-local interfaces.
+	ip-transparent: no
```
lots of other good options here

`harden-glue`, `harden-dnssec-stripped`, `edns-buffer-size`, `use-caps-for-id`, `prefetch`, etc

alexvojproc replied:
`harden-glue`, `harden-dnssec-stripped` and `harden-short-bufsize` (edns) are enabled by default. will have to discuss `prefetch`

alexvojproc committed:
chore(securedns): first review changes

HastD commented on files/system/usr/lib/systemd/system/dnsconfd.service.d/secureblue.conf:
```diff
@@ -0,0 +1,39 @@
+LockPersonality=yes
+RestrictNamespaces=yes
+PrivateDevices=yes
+SystemCallArchitectures=native
```
Did you mean to delete the syscall filter entirely, rather than just leaving it as `SystemCallFilter=@system-service` as I suggested?

alexvojproc replied:
Oops, I didn't mean to do that. Apologies.

alexvojproc commented:
Changes:
- Hopefully all review points addressed
- Added integration test, including for /usr/bin/dnsconfd SELinux type
- Reverted SELinux stub policy in favour of dnsconfd v1.7.3 which was recently pushed to F42 stable
- Use files module for resolv.conf installation to overwrite symlink
- General style, consistency, clarificatory comments

This PR should be ready (pending re-review) but if [v1.7.3](https://bodhi.fedoraproject.org/updates/FEDORA-2025-2eb1c80e01) hasn't hit mirrors by the time it's merged, then the integration test will fail.
